### PR TITLE
Inclusion of backbone models for general purpose tasks

### PR DESCRIPTION
**Pull request by PabloAMC**
Code adapted mostly from benchmark folder to create models not so specific as the present ones

### CODE REVIEW

**Comment by rusty1s** [comment]
Thanks a lot. Before this becomes too complicated to review, can we first merge some of the basic GNNs without any fancy pooling?

**Comment by PabloAMC** [comment]
I have added models for GINE, GATv2, ASAPool, SAGPool, Graclus and EdgePool, including tests

File: torch_geometric/nn/models/basic_gnn.py [hunk 1]
```diff
@@ -192,6 +192,50 @@ def MLP(in_channels: int, out_channels: int) -> torch.nn.Module:
         )
 
 
+class GINE(BasicGNN):
```
Comment (rusty1s) [inline torch_geometric/nn/models/basic_gnn.py:195]: We can also inherit from `GIN`, which makes the `GIN.MLP` call a little bit more cleaner.

File: torch_geometric/nn/models/basic_gnn.py [hunk 2]
```diff
@@ -192,6 +192,50 @@ def MLP(in_channels: int, out_channels: int) -> torch.nn.Module:
...
+    r"""The Graph Neural Network from the `"How Powerful are Graph Neural
+    Networks?" <https://arxiv.org/abs/1810.00826>`_ paper, using the
+    :class:`~torch_geometric.nn.GINConv` operator for message passing.
```
Comment (rusty1s) [inline torch_geometric/nn/models/basic_gnn.py:198]: GINE

File: test/nn/models/test_basic_gnn.py [hunk 3]
```diff
@@ -52,6 +52,19 @@ def test_gin(dropout, act, norm, jk):
     assert model(x, edge_index).size() == (3, out_channels)
 
 
+@pytest.mark.parametrize('dropout,act,norm,jk',
+                         product(dropouts, acts, norms, jks))
+def test_gine(dropout, act, norm, jk):
```
Comment (rusty1s) [inline test/nn/models/test_basic_gnn.py:57]: The `GINE` module expects `edge_attr` as input to the `forward` function.

File: torch_geometric/nn/models/basic_gnn.py [hunk 4]
```diff
@@ -192,6 +192,50 @@ def MLP(in_channels: int, out_channels: int) -> torch.nn.Module:
...
+        dropout (float, optional): Dropout probability. (default: :obj:`0.`)
+        act (Callable, optional): The non-linear activation function to use.
+            (default: :meth:`torch.nn.ReLU(inplace=True)`)
+        norm (torch.nn.Module, optional): The normalization operator to use.
+            (default: :obj:`None`)
+        jk (str, optional): The Jumping Knowledge mode
+            (:obj:`"last"`, :obj:`"cat"`, :obj:`"max"`, :obj:`"last"`).
+            (default: :obj:`"last"`)
+        **kwargs (optional): Additional arguments of
+            :class:`torch_geometric.nn.conv.GINConv`.
```
Comment (rusty1s) [inline torch_geometric/nn/models/basic_gnn.py:213]: GINE

File: torch_geometric/nn/models/basic_gnn.py [hunk 5]
```diff
@@ -231,3 +275,44 @@ def __init__(self, in_channels: int, hidden_channels: int, num_layers: int,
             GATConv(in_channels, out_channels, dropout=dropout, **kwargs))
         for _ in range(1, num_layers):
             self.convs.append(GATConv(hidden_channels, out_channels, **kwargs))
+
+
+class GATv2(BasicGNN):
+    r"""The Graph Neural Network from the `"Graph Attention Networks"
+    <https://arxiv.org/abs/1710.10903>`_ paper, using the
+    :class:`~torch_geometric.nn.GATConv` operator for message passing.
```
Comment (rusty1s) [inline torch_geometric/nn/models/basic_gnn.py:283]: GATv2

File: torch_geometric/nn/models/basic_gnn.py [hunk 6]
```diff
@@ -231,3 +275,44 @@ def __init__(self, in_channels: int, hidden_channels: int, num_layers: int,
...
+        num_layers (int): Number of GNN layers.
+        dropout (float, optional): Dropout probability. (default: :obj:`0.`)
+        act (Callable, optional): The non-linear activation function to use.
+            (default: :meth:`torch.nn.ReLU(inplace=True)`)
+        norm (torch.nn.Module, optional): The normalization operator to use.
+            (default: :obj:`None`)
+        jk (str, optional): The Jumping Knowledge mode
+            (:obj:`"last"`, :obj:`"cat"`, :obj:`"max"`, :obj:`"last"`).
+            (default: :obj:`"last"`)
+        **kwargs (optional): Additional arguments of
```
Comment (rusty1s) [inline torch_geometric/nn/models/basic_gnn.py:297]: GATv2

File: torch_geometric/nn/models/basic_pools.py [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+from torch.nn import ModuleList, Sequential, Linear, BatchNorm1d, ReLU
+
+from torch_geometric.data import Batch
+from torch_geometric.nn.conv import GCNConv, SAGEConv, GINConv, GINEConv, GATConv, GATv2Conv, GraphConv
+from torch_geometric.nn.pool import EdgePooling, SAGPooling, ASAPooling, graclus, max_pool
+from torch_geometric.nn.glob import global_mean_pool, global_max_pool, global_add_pool
+from torch_geometric.nn.models.jumping_knowledge import JumpingKnowledge
+
+
+class BasicPoolGNN(torch.nn.Module):
```
Comment (rusty1s) [inline torch_geometric/nn/models/basic_pools.py:18]: Is it possible to inherit from `BasicGNN`?

File: torch_geometric/nn/models/basic_pools.py [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+            (default: :obj:`"last"`)
+        glob_pool (str, optional): The global pooling operator to use
+            (:obj:`"add"`, :obj:`"max"`, :obj:`"mean"`).
+            (default: :obj:`"mean"`)
+    """
+    def __init__(self, in_channels: int, hidden_channels: int, num_layers: int,
+                 dropout: float = 0.0,
+                 act: Optional[Callable] = ReLU(inplace=True),
+                 norm: Optional[torch.nn.Module] = None, jk: str = 'last',
+                 glob_pool: str = 'mean'):
```
Comment (rusty1s) [inline torch_geometric/nn/models/basic_pools.py:41]: `global_pool`?

File: torch_geometric/nn/models/basic_pools.py [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+        for i in range(self.num_layers):
+            x = self.convs[i](x, edge_index, *args, **kwargs)
+            if self.norms is not None:
+                x = self.norms[i](x)
+            if self.act is not None:
+                x = self.act(x) 
+
+            xs += [self.global_pool(x, batch)]
+            if i % 2 == 0 and i < len(self.convs) - 1:
+                pool = self.pools[i // 2]
```
Comment (rusty1s) [inline torch_geometric/nn/models/basic_pools.py:90]: I think we should create a function `self.pool(layer, x, edge_index, batch, *kwargs)` that takes care of pooling, and that one needs to override. I think not all pooling ops are perfectly aligned to this interface yet.

File: torch_geometric/nn/models/basic_pools.py [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+                 dropout: float = 0.0,
+                 act: Optional[Callable] = ReLU(inplace=True),
+                 norm: Optional[torch.nn.Module] = None, jk: str = 'last',
+                 glob_pool: str = 'mean',
+                 ratio: float = 0.8,
+                 **kwargs):
+        super().__init__(in_channels, hidden_channels, num_layers, dropout,
+                         act, norm, jk, glob_pool)
+
+        self.convs.append(GraphConv(in_channels, hidden_channels, **kwargs))
```
Comment (rusty1s) [inline torch_geometric/nn/models/basic_pools.py:140]: Can we allow  different GNN ops as well?

**Review by rusty1s — CHANGES_REQUESTED** [review]

**Comment by PabloAMC** [comment]
I'm closing this PR because the basic models were already open.

**Pull request closed without merging**